db/db_impl.cc

```diff
@@ -4035,17 +4037,28 @@ Status DBImpl::SwitchMemtable(ColumnFamilyData* cfd, WriteContext* context) {
     if (creating_new_log) {
       EnvOptions opt_env_opt =
           env_->OptimizeForLogWrite(env_options_, db_options_);
-      s = env_->NewWritableFile(
-          LogFileName(db_options_.wal_dir, new_log_number), &lfile,
-          opt_env_opt);
+      if (recycle_log_number) {
+        Log(InfoLogLevel::INFO_LEVEL, db_options_.info_log,
+            "reusing log %" PRIu64 " from recycle list\n", recycle_log_number);
+        s = env_->ReuseWritableFile(
+            LogFileName(db_options_.wal_dir, new_log_number),
+            LogFileName(db_options_.wal_dir, recycle_log_number), &lfile,
+            opt_env_opt);
+      } else {
+        s = env_->NewWritableFile(
+            LogFileName(db_options_.wal_dir, new_log_number), &lfile,
+            opt_env_opt);
+      }
       if (s.ok()) {
         // Our final size should be less than write_buffer_size
         // (compression, etc) but err on the side of caution.
-        lfile->SetPreallocationBlockSize(
-            1.1 * mutable_cf_options.write_buffer_size);
+        lfile->SetPreallocationBlockSize(1.1 *
+                                         mutable_cf_options.write_buffer_size);
```

Comment:
Just curious about the implication of it. It will pretty much make we call fallocate(0, <this_size>) in the beginning of the file. Would it slow down writes for allocating more space or zeroing out?

To make things more complicated, we have this: https://github.com/facebook/rocksdb/blob/master/util/env_posix.cc#L1277-L1280 @igorcanadi is it still a problem?

Just a question to discussion. I'm OK with commit as it is no matter what the behavior is. We can always optimize later if there is an opportunity.